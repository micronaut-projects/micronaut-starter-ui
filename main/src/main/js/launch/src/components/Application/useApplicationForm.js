import { useEffect, useMemo, useRef, useState } from 'react'
import { MicronautStarterSDK } from '../../micronaut'

import {
  resetRoute,
  resolveActionRoute,
  isDeepLinkReferral,
  ACTIVITY_KEY,
  parseAndConsumeQuery,
  sharableLink,
} from '../../helpers/Routing'
import useLocalStorage from '../../hooks/useLocalStorage'
import Lang from '../../helpers/Lang'

export function useShareData() {
  return useRef(parseAndConsumeQuery())
}

function formResets(fallbacks = {}) {
  const { name, package: pkg, type } = fallbacks
  return {
    name: typeof name === 'string' ? name : 'demo',
    package: typeof pkg === 'string' ? pkg : 'com.example',
    type: typeof type === 'string' ? type : 'DEFAULT',
  }
}

const initialForm = (initialData) => {
  const { javaVersion, lang, build, test } = initialData
  const parsed = {
    javaVersion: typeof javaVersion === 'string' ? javaVersion : '', // This is specifically "" to work with the SelectOption component
    lang: typeof lang === 'string' ? lang : null,
    build: typeof build === 'string' ? build : null,
    test: typeof test === 'string' ? test : null,
  }
  return {
    ...parsed,
    ...formResets(initialData),
  }
}

export function useInitialVersionEffect(
  initialVersion,
  availableVersions = [],
  setSelectedVersion,
  handleWarning
) {
  useEffect(() => {
    const hasVersions =
      Array.isArray(availableVersions) && availableVersions.length > 0

    const checkRequestedVersion = (hasVersions, versions) => {
      if (hasVersions && initialVersion) {
        const match = versions.find((v) => v.version === initialVersion)
        if (!match) {
          const message = Lang.trans('error.versionNoLongerSupported', {
            requestedVersion: initialVersion,
            currentVersion: versions[0].version,
          })
          handleWarning(message)
        }
      }
    }

    checkRequestedVersion(hasVersions, availableVersions)

    hasVersions &&
      setSelectedVersion((version) => {
        return (
          availableVersions.find((v) => v.version === version?.version) ||
          availableVersions.find((v) => v.version === initialVersion) ||
          availableVersions[0]
        )
      })
  }, [availableVersions, initialVersion, setSelectedVersion, handleWarning])
}

export function useFeaturesHandler(setFeaturesSelected) {
  return useMemo(() => {
    const addFeature = (feature) => {
      setFeaturesSelected(({ ...draft }) => {
        draft[feature.name] = feature
        return draft
      })
    }

    const removeFeature = (feature) => {
      setFeaturesSelected(({ ...draft }) => {
        delete draft[feature.name]
        return draft
      })
    }

    const removeAllFeatures = () => {
      setFeaturesSelected({})
    }

    return {
      addFeature,
      removeFeature,
      removeAllFeatures,
    }
  }, [setFeaturesSelected])
}

export function useOnInitialLoadEffect(initialData, onRepoCreated, onError) {
  // Use Effect Hook For Error Handling and
  // GitHub on complete callback
  useEffect(() => {
    const { error, htmlUrl, cloneUrl } = initialData
    if (!error && !htmlUrl) {
      return // nothing more to do
    }
    resetRoute()
    setTimeout(() => {
      if (cloneUrl) {
        onRepoCreated({
          cloneUrl,
          htmlUrl,
          show: true,
          type: 'clone',
        })
      } else if (error) {
        onError(error.replaceAll('+', ' '))
      }
    }, 500)
  }, [initialData, onRepoCreated, onError])
}

export function useHandleShareLinkEffect(
  sdk,
  initialData,
  routingHandlers = {}
) {
  const isReferral = isDeepLinkReferral(initialData)
  const shareData = useRef(initialData)

  // Deep Linking
  useEffect(() => {
    async function handleDeepLink() {
      const activity = resolveActionRoute(shareData.current)
      if (!activity) {
        return
      }

      // If we're not able to handle the route,
      // discard and reset the history state
      if (!isReferral || !Object.keys(routingHandlers).includes(activity)) {
        // Push back and return
        resetRoute()
        return
      }

      const { showing } = shareData.current

      delete shareData.current[ACTIVITY_KEY]
      delete shareData.current.showing

      // This is a common react problem
      // Since we have to wait for the SDK to get initialized
      // but can't watch the create / features objects,
      // we need to rebuild at routing time.
      const payload = {
        ...initialForm(shareData.current),
        features: MicronautStarterSDK.reconstructFeatures(
          shareData.current.features
        ),
      }

      await routingHandlers[activity](payload, sdk, { showing })
    }

    if (sdk?.baseUrl) {
      handleDeepLink()
    }
  }, [sdk, routingHandlers, isReferral, shareData])
}

const EMPTY_DATA = {}

export function useApplicationForm(initialData = EMPTY_DATA) {
  // Form Data
  const isReferral = isDeepLinkReferral(initialData)

  const [form, setForm] = useLocalStorage(
    'INITIAL_FORM_DATA',
    initialForm(initialData),
    isReferral
  )

  const [selectedFeatures, setSelectedFeatures] = useState(
    MicronautStarterSDK.reconstructFeatures(initialData.features)
  )

  const [selectedVersion, setSelectedVersion] = useLocalStorage(
    'SELECTED_MN_VERSION',
    null,
    isReferral
  )

  const resetHandlers = useMemo(() => {
    const onResetFeatures = () => {
      setSelectedFeatures({})
    }

    const onResetForm = () => {
      setForm((form) => ({ ...form, ...formResets() }))
    }

    const onResetAll = () => {
      onResetForm()
      onResetFeatures()
    }

    return { onResetForm, onResetFeatures, onResetAll }
  }, [setForm, setSelectedFeatures])

  const createPayload = { ...form, features: selectedFeatures }

  const sharable = useMemo(() => {
    return sharableLink(form, selectedFeatures, selectedVersion?.version)
  }, [selectedFeatures, form, selectedVersion?.version])

  return {
    sharable,
    createPayload,
    form,
    setForm,
    selectedFeatures,
    setSelectedFeatures,
    selectedVersion,
    setSelectedVersion,
    resetHandlers,
  }
}
