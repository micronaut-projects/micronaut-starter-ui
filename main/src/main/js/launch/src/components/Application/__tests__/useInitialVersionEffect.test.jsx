// Link.react.test.js
import React, { useState } from 'react'
import { create, act } from 'react-test-renderer'

import { useInitialVersionEffect } from '../useApplicationForm'

const TestView = ({ initialVersion, availableVersions, handlerWarning }) => {
  const [version, setSelectedVersion] = useState(null)

  useInitialVersionEffect(
    initialVersion,
    availableVersions,
    setSelectedVersion,
    handlerWarning
  )
  return <div className="test">{version?.version}</div>
}

const TEST_DATA = [
  { initialVersion: undefined, expected: '2.1', hasError: false },
  { initialVersion: null, expected: '2.1', hasError: false },
  { initialVersion: '2.1', expected: '2.1', hasError: false },
  { initialVersion: '2.0', expected: '2.1', hasError: true },
  { initialVersion: '2.2-SNAPSHOT', expected: '2.2-SNAPSHOT', hasError: false },
]

TEST_DATA.forEach(({ initialVersion, expected, hasError }) => {
  it(`Initial Version with value of "${initialVersion}"`, () => {
    let availableVersions = [{ version: '2.1' }, { version: '2.2-SNAPSHOT' }]

    let error = null
    const handlerWarning = (e) => {
      error = e
    }

    let testRenderer
    act(() => {
      testRenderer = create(
        <TestView
          initialVersion={initialVersion}
          availableVersions={availableVersions}
          handlerWarning={handlerWarning}
        />
      )
    })

    const testInstance = testRenderer.root
    expect(testRenderer.toJSON()).toMatchSnapshot()
    expect(testInstance.findByProps({ className: 'test' }).children).toEqual([
      expected,
    ])

    if (hasError) {
      expect(error).toEqual(
        `Micronaut Launch no longer supports version ${initialVersion} we're setting the configuration to the most recent release version ${availableVersions[0].version}`
      )
    }
  })
})
