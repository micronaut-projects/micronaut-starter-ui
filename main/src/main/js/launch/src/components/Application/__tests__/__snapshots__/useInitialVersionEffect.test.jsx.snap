// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Version with value of "2.0" 1`] = `
<div
  className="test"
>
  2.1
</div>
`;

exports[`Initial Version with value of "2.1" 1`] = `
<div
  className="test"
>
  2.1
</div>
`;

exports[`Initial Version with value of "2.2-SNAPSHOT" 1`] = `
<div
  className="test"
>
  2.2-SNAPSHOT
</div>
`;

exports[`Initial Version with value of "null" 1`] = `
<div
  className="test"
>
  2.1
</div>
`;

exports[`Initial Version with value of "undefined" 1`] = `
<div
  className="test"
>
  2.1
</div>
`;
